spatialprofilingtoolbox/db/feature_matrix_extractor.py

```diff
@@ -16,42 +19,67 @@
 logger = colorized_logger(__name__)
 
 
+class DBSource(Enum):
+    """Indicator of intended database source."""
+    CURSOR = auto()
+    CONFIG_FILE = auto()
+    UNKNOWN = auto()
+
+
 class FeatureMatrixExtractor:
     """
     Pull from the database and create convenience bundle of feature matrices and metadata.
     """
 
-    cursor: Psycopg2Cursor | None
+    cursor: Psycopg2Cursor
     database_config_file: str | None
+    db_source: DBSource
 
     def __init__(self,
         cursor: Psycopg2Cursor | None=None,
         database_config_file: str | None=None,
     ):
-        if cursor is None and database_config_file is None:
+        self.cursor = cast(Psycopg2Cursor, cursor)
+        self.database_config_file = database_config_file
+        if cursor is not None:
+            self.db_source = DBSource.CURSOR
+        elif database_config_file is not None:
+            self.db_source = DBSource.CONFIG_FILE
+        else:
+            self.db_source = DBSource.UNKNOWN
+        self._report_on_arguments()
+
+    def _report_on_arguments(self):
+        if self.cursor is None and self.database_config_file is None:
             logger.error('Must supply either cursor or database_config_file.')
-            return
-        if cursor is not None and database_config_file is not None:
+        if self.cursor is not None and self.database_config_file is not None:
             message = 'A cursor and database configuration file were both specified. Using the '\
                 'cursor.'
             logger.warning(message)
-        self.cursor = cursor
-        self.database_config_file = database_config_file
 
     def extract(self,
         specimen: str | None=None,
         study: str | None=None,
         continuous_also: bool=False,
     ):
         extraction = None
-        if self.cursor is not None:
-            extraction = self._extract(specimen=specimen, study=study, continuous_also=continuous_also)
-        if self.database_config_file is not None:
+        if self.db_source == DBSource.CURSOR:
+            extraction = self._extract(
+                specimen=specimen,
+                study=study,
+                continuous_also=continuous_also,
+            )
+        if self.db_source == DBSource.CONFIG_FILE:
```

Comment:
Not a blocker due to the new way it's implemented, but this should still be an `elif`, right?

Reply:
I'm not sure, I've seen it recommended to use a sequence of bare `if` statements when exhaustively handling an enum value. I guess for symmetry.
The behavior is theoretically different, since the enum value (`db_source` in this case) could accidentally be altered during one of the handlers...

Reply:
(Basically, Python lacks `switch`).

Reply:
Wait it doesn't. I just saw [this](https://docs.python.org/3.10/whatsnew/3.10.html#pep-634-structural-pattern-matching), apparently availabile since 3.10.

Reply:
Updated to use `match case`.